servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyPipelinedConnection.java

```diff
@@ -0,0 +1,628 @@
+/**
+ * Contract for using a {@link NettyConnection} to make pipelined requests, typically for a client.
+ * <p>
+ * Pipelining allows to have concurrent requests processed on the server but still deliver responses in order.
+ * This eliminates the need for request-response correlation, at the cost of head-of-line blocking.
+ * @param <Req> Type of requests sent on this connection.
+ * @param <Resp> Type of responses read from this connection.
+ */
+final class NettyPipelinedConnection<Req, Resp> implements NettyConnectionContext {
+    private static final Logger LOGGER = LoggerFactory.getLogger(NettyPipelinedConnection.class);
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<NettyPipelinedConnection, Node>
+            writeQueueTailUpdater = newUpdater(NettyPipelinedConnection.class, Node.class, "writeQueueTail");
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<NettyPipelinedConnection, Node>
+            readQueueTailUpdater = newUpdater(NettyPipelinedConnection.class, Node.class, "readQueueTail");
+    @SuppressWarnings("unused")
+    @Nullable
+    private volatile Node writeQueueTail;
+    @SuppressWarnings("unused")
+    @Nullable
+    private volatile Node readQueueTail;
+
+    private final NettyConnection<Resp, Req> connection;
+
+    /**
+     * New instance.
+     *
+     * @param connection {@link NettyConnection} requests to which are to be pipelined.
+     */
+    NettyPipelinedConnection(NettyConnection<Resp, Req> connection) {
+        this.connection = requireNonNull(connection);
+    }
+
+    /**
+     * Do a write operation in a pipelined fashion.
+     * @param requestPublisher {@link Publisher} producing the request(s) to write.
+     * @param flushStrategySupplier The {@link FlushStrategy} to use for this write operation.
+     * @param writeDemandEstimatorSupplier A {@link Supplier} of {@link WriteDemandEstimator} for this request which
+     * impacts how many elements are requested from the {@code requestPublisher} depending upon channel writability.
+     * @return Response {@link Publisher} for this request.
+     */
+    Publisher<Resp> write(final Publisher<Req> requestPublisher,
+                          final Supplier<FlushStrategy> flushStrategySupplier,
+                          final Supplier<WriteDemandEstimator> writeDemandEstimatorSupplier) {
+        return Publisher.defer(() -> {
+            // Lazy modification of local state required (e.g. nodes, delayed subscriber, queue modifications)
+
+            // Setup read side publisher and nodes
+            DelayedSubscribePublisher<Resp> delayedResponsePublisher = new DelayedSubscribePublisher<>(
+                    toSource(connection.read()));
+            Node readNode = new Node(delayedResponsePublisher::processSubscribers);
+            Publisher<Resp> composedResponsePublisher =
+                    delayedResponsePublisher.liftSync(new ReadPopNextOperator(readNode));
```

Comment:
```java
    /**
     * Logically equivalent to {@link Publisher#afterFinally(Runnable)} but relies upon internal queue CAS operations
     * to prevent multiple executions (e.g. reduces a CAS operation).
     */
```